### ISSUE TITLE
Enforce nesting with ampersand

### ISSUE THREAD

**Issue by stephenway** [stylelint/stylelint#489]
Per the nesting spec you are always supposed to specify the nesting selector. Having an option to enforce this would be helpful to write more closely to the spec.
http://tabatkins.github.io/specs/css-nesting/

**Comment by MadLittleMods** [comment]
:+1:

> A style rule can be directly nested within another style rule if its selector is nest-prefixed.
> http://tabatkins.github.io/specs/css-nesting/#direct
> ```
> .foo {
>   color: blue;
>   & > .bar { color: red; }
> }
> ```

**Comment by davidtheclark** [comment]
This looks like it would be a pretty simple rule to write. Are these all the requirements?

- For any nested rule, ensure there is `&` within the first compound selector.
- For any `@nest` at-rule, ensure there is a `&` anywhere within the selector.

**Comment by stephenway** [comment]
@davidtheclark That seems to cover it. :+1:

**Comment by tbremer** [comment]
Current [nesting block rules](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/rules.md#nesting-block) look as if a curly brace is required, which is not in the spec… 

How would additional rules around nesting affect these?

What about nesting and [type selection](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/rules.md#selector)? This is currently considered an error if `selector-no-type` is turned on and you are nesting per the spec:
```css
.foo {
    color: orange;

    & [class*=bar] {
        color: blue
    }
}
```

I am more than happy to write up a rule… Would like a bit of clarification.

**Comment by tbremer** [comment]
@davidtheclark The check will also need to look for an ampersand within lists of selectors as well:
```css
.foo {
  color: red;
  @nest & .bar, .baz {
    color: blue;
  }
}
/* Invalid because not all selectors in the list
   contain a nesting selector */
```

**Comment by stephenway** [comment]
@tbremer Yes we will want to check each selector in a list for `&`. Good catch!

**Comment by jeddy3** [comment]
Thanks @tbremer, @stephenway and @MadLittleMods for the input.

I looks like there are three things being discussed in this thread. 

1. A new rule to check for invalid css-nesting syntax.
2. The validity of the `nesting-block-opening-brace-newline/space-before` rules.
3. The `selector-no-type` rule doesn't exclude the _nesting selector (&)_.

I've broken the latter two into their own issues: #498 & #499

I think we can capture this in one rule, perhaps called `rule-no-invalid-nesting`. Although, the _thing_ this rule should be attached to isn't really obvious. 

> @tbremer Yes we will want to check each selector in a list for &. Good catch!

So, @davidtheclark specs updated look something like:

- For any direct nested rule, ensure that for each _selector_ within the _selector list_ `&` is the _first simple selector_ of the _first compound selector_.
- For any `@nest` at-rule, ensure that for each _selector_ within the _selector list_ a `&` is anywhere.

Sound about right?

@davidtheclark What are your thoughts about handling this is a plugin? It'll save us a breaking change should the spec change again.

> @tbremer I am more than happy to write up a rule…

That's fantastic! Lets wait until we decide on whether this is `rule` or `plugin` material and then feel free to take a crack at it :)

**Comment by davidtheclark** [comment]
> What are your thoughts about handling this is a plugin?

I think that's a good idea. we were premature with those other two rules, so let's avoid that in the future. @tbremer if you'd like to make the plugin, that would be great!

**Comment by jeddy3** [comment]
@tbremer Actually, you might find it more manageable to break this into a couple of different rules:

* `rule-directly-nested-no-missing-nest-prefixed`
* `rule-at-nest-no-missing-nest-containing`

This will then lend itself nicely to future nesting rules e.g. 

1. Say you wish, for the sake of consistency, to disallow _directly nested_ and only use `at-nest` ones, you can create `rule-no-directly-nested`.
2. Or you thought `.class &` is an anti-pattern and want to restrict nesting to what's offered by _directly nesting_ then you could create `rule-no-at-nest`.

In https://github.com/stylelint/stylelint/issues/497 we're talking about bundling plugins, so you'll be able to group everything together into something like `stylelint-nesting`.

Just food for thought :)

**Comment by tbremer** [comment]
I read the plugin developer guides. When do you choose a plugin over a ruleset?

It seems to me with nesting going into the spec this should become a rule?

**Comment by jeddy3** [comment]
> When do you choose a plugin over a set of rules?

That's a good question. We've not really nailed down a hard and fast distinction yet.

Originally we decided on rules being for widely applicable _standard_ CSS and plugins for _non standard syntax_ or very specific use cases. I think we're now thinking, since prematurely writing the `nesting-block-*` rules, that the former should be become "widely applicable _standard_ CSS that is unlikely to change any time soon". Does that sound about right @davidtheclark?

> but I don't know the motivations

There was a [discussion back in May](https://github.com/stylelint/stylelint/issues/77#issuecomment-103924979) about keeping stuff that might change or become redundant out of core. In that instance the discussion was about supporting a particular methodology, but the question is the same; what do we focus our finite efforts on? By making something a rule we're saying we, the stylelint team, will maintain it.  And so I think we settled on the rules that'll appeal to the most number of people and that are least likely to change over time, and then create a plugin system so people can have a go at, and support, anything outside of this.

> Also, are you in general against changing the meaning or syntax of a rule? Looks like #498 brings that point up though…

I'm not 100% sure what you mean, so forgive me if I've gotten the wrong end of the stick :) If I have btw, do you mind expanding a little on your original question?

I'm not very familiar with the nesting spec, but [it looks like](http://discourse.wicg.io/t/css-nesting-specification/839/17) the _thing_ called a _nesting block_ doesn't exist anymore. So, I'm not sure there's much scope to change the meaning or syntax of this rule, when the _thing_ that the rule is named after has gone.  

Perhaps there's a good argument here for naming things a little more generally e.g. `stylelint-rule-no-invalid-nesting`, so if the details of nesting change again you can adapt to the changes in meaning or syntax without scraping the rule. Although, we've found that having narrowly focused rules that check one thing only has been quite helpful when writing them.

> It seems to me with nesting going into the spec this should become a rule.

So, to summarise the above... I think we now also need to think about about the maturity of the spec when deciding if it's a plugin or rule. Perhaps the nesting stuff can start out as a plugin and then become a rule once things stabilise a little and more people adopt it.

I hope that helps. Like I said at the beginning, we've kind of played this one by ear so far. So, any input is more than welcome :)

**Comment by davidtheclark** [comment]
If we could make a rule that would apply both to Sass syntax and to the CSS spec draft, I think that could go into core. I know that when I was using Sass I tried to be explicit about my nested selectors and use `&`. But if the rule's just tailored to the unstable spec draft, let's keep it in a plugin for now. 

Maybe what could work for both Sass and future CSS is if the rule just enforced that each nested selector contain a `&` somewhere within it. A more specialized plugin could enforce that the `&` be at the *beginning* of spec-draft-style-directly-nested selectors. Same goes for rules about at-nest.

How does that sound?

**Comment by jeddy3** [comment]
> Maybe what could work for both Sass and future CSS is if the rule just enforced that each nested selector contain a & somewhere within it.

Sgtm. Maybe called `rule-nesting-no-missing-ampersand`, as to keep with the precedence set by [`no-missing-eof-newline`](https://github.com/stylelint/stylelint/blob/master/src/rules/no-missing-eof-newline/README.md)?

Then, the likes of:

* `stylelint-rule-no-invalid-directly-nested` - checks for valid _nested-prefixed_
* `stylelint-rule-no-invalid-at-nest` - checks for valid _nested-containing_

etc.

As plugins for the time being.

**Comment by davidtheclark** [comment]
Sounds good. Should we make a new issue for rule-nesting-no-missing-ampersand?

@stephenway, @MadLittleMods, @tbremer : What do you think of that plan?

**Comment by stephenway** [comment]
@davidtheclark Sounds right, I like it!

**Comment by tbremer** [comment]
The problem in supporting both Sass and CSS in the same rule is that this is valid in Sass and totally invalid in CSS. *For the example below I am assuming the SCSS Syntax of Sass:*
```scss
/**
 * this is valid Sass, and in no way valid CSS based on the spec
*/
a, b {
   color: orange

    c, &d {
        color: blue;
    }
}
```

I like, as @jeddy3 said keeping rules that are not integrated in the spec out of the core, so something like this could / should be a plugin until the point where W3 takes up talking about the nested spec / it moves  past it's current stage

**Comment by jeddy3** [comment]
> The problem or difficulty in supporting both Sass and CSS in the same rule is that this is valid in Sass and totally invalid in CSS

I think @davidtheclark was referring to a specific self-imposed rule when using nesting with sass i.e. always using an `&` for each selector in a list, even though it is optional in sass for descendent selectors.  

>  so something like this could / should be a plugin until the point 

I think you're right. @davidtheclark Soz to flip-flop on this but shall we push all nesting stuff to plugins for now?

So, @tbremer, either way feel free to start on the likes of `stylelint-rule-no-invalid-directly-nested` and `stylelint-rule-no-invalid-at-nest` as it sounds like you guys already have a need for those. 

And shall we park `rule-nesting-no-missing-ampersand` as no one has expressed a need for it and it'll likely be made redundant going forward as the spec-based nesting is adopted? And if someone does need to enforce the aforementioned self-imposed nesting rule, then they can knock up their own plugin for it.

Sound groovy?

**Comment by davidtheclark** [comment]
Ok.
